src/browser/cdp/selectivity/utils.ts

```diff
@@ -0,0 +1,210 @@
+        }
+    }
+
+    return dependantSourceFiles;
+};
+
+export const hasProtocol = (fileUrlLikePath: string): boolean => {
+    if (!fileUrlLikePath.includes("://")) {
+        return false;
+    }
+
+    try {
+        return Boolean(new URL(fileUrlLikePath).protocol);
+    } catch {
+        return false;
+    }
+};
+
+const ensurePosixRelativeDependencyPathExists = memoize((posixRelativePath: string): void => {
+    const relativePath = posixRelativePath.replaceAll(path.posix.sep, path.sep);
+
+    if (fs.existsSync(relativePath)) {
+        return;
+    }
+
+    throw new Error(`Selectivity: Couldn;t find "${relativePath}", which is test's dependency`);
+});
+
+/**
+ * @param cssDependencies SORTED uniq array of css dependenciy URI's
+ * @param jsDependencies SORTED uniq array of js dependenciy URI's
+ * @returns sorted uniq arrays of relative paths
+ */
+export const transformSourceDependencies = (
+    cssDependencies: string[],
+    jsDependencies: string[],
+): NormalizedDependencies => {
+    const nodeModulesLabel = "node_modules/";
+    const css: string[] = [];
+    const js: string[] = [];
+    const modules: string[] = [];
+
+    const classifyDependency = (dependency: string, typedResultArray: string[]): void => {
+        dependency = decodeURIComponent(softFileURLToPath(dependency));
+
+        if (hasProtocol(dependency)) {
+            throw new Error(`Selectivity: Found unsupported protocol in dependencies ("${dependency}")`);
+        }
+
+        const dependencyRelativePath = path.posix.relative(path.posix.resolve(), path.posix.resolve(dependency));
+
+        const nodeModulesLabelPos = dependencyRelativePath.indexOf(nodeModulesLabel);
+
+        if (nodeModulesLabelPos === -1) {
+            ensurePosixRelativeDependencyPathExists(dependencyRelativePath);
+            typedResultArray.push(dependencyRelativePath);
+            return;
+        }
+
+        const modulePos = nodeModulesLabelPos + nodeModulesLabel.length;
+        const isScopedDependency = dependencyRelativePath[modulePos] === "@";
+        let moduleEndPos;
+
+        if (isScopedDependency) {
+            const scopeEndPos = dependencyRelativePath.indexOf("/", modulePos + 1);
+
+            moduleEndPos = scopeEndPos === -1 ? -1 : dependencyRelativePath.indexOf("/", scopeEndPos + 1);
+        } else {
+            moduleEndPos = dependencyRelativePath.indexOf("/", modulePos + 1);
+        }
+
+        if (moduleEndPos === -1) {
+            ensurePosixRelativeDependencyPathExists(dependencyRelativePath);
+            typedResultArray.push(dependencyRelativePath);
+        } else {
+            const modulePath = dependencyRelativePath.slice(0, moduleEndPos);
+
+            if (modules[modules.length - 1] !== modulePath) {
+                ensurePosixRelativeDependencyPathExists(modulePath);
+
+                modules.push(modulePath);
+            }
+        }
+    };
+
+    let cssIndex = 0;
+    let jsIndex = 0;
```

Comment:
using 2 pointers to create merged "modules" array of uniq modules from sorted "css" and "js" arrays